identity_stardust/src/did/stardust_did.rs

```diff
@@ -0,0 +1,927 @@
+  fn check_tag_str(tag: &str) -> Result<()> {
+    prefix_hex::decode::<[u8; TAG_BYTES_LEN]>(tag)
+      .map_err(|_| DIDError::InvalidMethodId)
+      .map(|_| ())
+  }
+
+  // Normalizes the DID `method_id` by removing the default network segment if present.
+  //
+  // E.g.
+  // - `"did:stardust:main:123" -> "did:stardust:123"` is normalized
+  // - `"did:stardust:dev:123" -> "did:stardust:dev:123"` is unchanged
+  // TODO: Remove the lint once this bug in clippy has been fixed. Without to_owned a mutable reference will be aliased.
+  #[allow(clippy::unnecessary_to_owned)]
+  fn normalize(mut did: CoreDID) -> CoreDID {
+    let method_id = did.method_id();
+    let (network, tag) = Self::denormalized_components(method_id);
+    if tag.len() == method_id.len() || network != NetworkName::DEFAULT_STR {
+      did
+    } else {
+      did
+        .set_method_id(tag.to_owned())
+        .expect("normalizing a valid CoreDID should be Ok");
+      did
+    }
+  }
+
+  /// Checks if the given `DID` has a valid [`StardustDID`] network name.
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input is not a valid network name according to the [`StardustDID`] method specification.
+  pub fn check_network<D: DID>(did: &D) -> Result<()> {
+    let network_name = Self::denormalized_components(did.method_id()).0;
+    NetworkName::validate_network_name(network_name).map_err(|_| DIDError::Other("invalid network name"))
+  }
+
+  /// Checks if the given `DID` is syntactically valid according to the [`StardustDID`] method specification.
+  ///
+  ///
+  ///  This function does NOT check whether the `did-method-specific-id` corresponds to a Blake2b-256 hash of an
+  ///  `OutputID`, despite this being a requirement for the [`StardustDID`] method specification.
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input is not a syntactically valid [`StardustDID`].
+  pub fn check_validity<D: DID>(did: &D) -> Result<()> {
+    Self::check_method(did).and_then(|_| Self::check_method_id(did))
+  }
+
+  /// Checks if the given `DID` has a valid [`StardustDID`] `method` (i.e. `"stardust"`).
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input represents another method.
+  // TODO: Change the naming in the docs once we remove the code for the current IOTA method.
+  pub fn check_method<D: DID>(did: &D) -> Result<()> {
+    (did.method() == Self::METHOD)
+      .then_some(())
+      .ok_or(DIDError::InvalidMethodName)
+  }
+
+  /// Checks if the given `DID` has a valid [`StardustDID`] `method_id`.
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input does not have a [`StardustDID`] compliant method id.
+  // TODO: Is it correct to also validate the network here? The current IOTA DID method does NOT do that.
+  pub fn check_method_id<D: DID>(did: &D) -> Result<()> {
+    let (network, tag) = Self::denormalized_components(did.method_id());
+    NetworkName::validate_network_name(network)
```

Comment:
That's probably an oversight in `IotaDID` that should be fixed. It doesn't affect internal usage because it's always constructed with a valid `NetworkName`. So the network check is technically redundant during construction, but not when called otherwise.

```suggestion
  pub fn check_method_id<D: DID>(did: &D) -> Result<()> {
    let (network, tag) = Self::denormalized_components(did.method_id());
    NetworkName::validate_network_name(network)
```

Reply:
Opened a new issue to fix the bug in `IotaDID`.